src/tensors/cpu/ruy_adapter.h

```diff
@@ -0,0 +1,534 @@
+
+      AlignedVector<std::int32_t> dst_data(rows_A * cols_B);
+      std::int32_t *dest_ptr = dst_data.data();
+
+      dst.set_data(dest_ptr);
+
+      // When Dst is int32, mul_params is unused.
+      ruy::MulParams<std::int32_t, std::int32_t> mul_params;
+      ruy::Mul(lhs, rhs, mul_params, &context, &dst);
+
+      // Unquantizes, then adds bias in a single statement on the output.
+      // float unquant_multiplier = (1.0f * scale_output) / (scale_A * scale_B);
+      Preprocess<kHighestPath>::unquantizeAddBias(
+          dest_ptr, bias_prepared, unquant_multiplier, rows_A, cols_B, output);
+    }
+
+    // output = A*B (notice no bias).
+    static void Multiply(const Type *input_A_prepared,
+                         const Type *input_B_prepared,
+                         float *output,
+                         Index rows_A,
+                         Index width,
+                         Index cols_B,
+                         float unquant_multiplier) {
+      // It is expected that somehow we have managed to call all prepare by the time
+      // we are here, with inputs (prepared) in int8_t. All that's left to do is use
+      // ruy for multiply and then start with the reverse ops to get to fp32.
+
+      // Use ruy to multiply.
+      // The following is adapted from
+      // https://github.com/google/ruy/blob/878283640de7946a43053e8ebf4f15114fbc9156/example/example.cc#L129-L152
+
+      ruy::Context context;
+      ruy::Matrix<std::int8_t> lhs;
+      ruy::MakeSimpleLayout(rows_A, width, ruy::Order::kRowMajor, lhs.mutable_layout());
+      lhs.set_data(input_A_prepared);
+
+      ruy::Matrix<std::int8_t> rhs;
+      ruy::MakeSimpleLayout(width, cols_B, ruy::Order::kColMajor, rhs.mutable_layout());
+      rhs.set_data(input_B_prepared);
+
+      ruy::Matrix<std::int32_t> dst;
+      ruy::MakeSimpleLayout(rows_A, cols_B, ruy::Order::kRowMajor, dst.mutable_layout());
+
+      AlignedVector<std::int32_t> dst_data(rows_A * cols_B);
+      std::int32_t *dest_ptr = dst_data.data();
+
+      dst.set_data(dest_ptr);
+
+      // When Dst is int32, mul_params is unused.
+      ruy::MulParams<std::int32_t, std::int32_t> mul_params;
+      ruy::Mul(lhs, rhs, mul_params, &context, &dst);
+
+      // Unquantizes, then adds bias in a single statement on the output.
+      // float unquant_multiplier = (1.0f * scale_output) / (scale_A * scale_B);
+      Preprocess<kHighestPath>::unquantize(dest_ptr, unquant_multiplier, rows_A, cols_B, output);
+    }
+  };
+
+  // Int16 support is currently missing.
+  struct Int16 : IntBase<int16_t> {
+    using Type = int16_t;
+  };
+
+  template <class T>
+  static T MaxAbsolute(const T *begin, const T *end) {
+    T result = 0;
+    for(auto p = begin; p < end; ++p) {
+      result = std::max(result, std::abs(*p));
+    }
+    return result;
+  }
+
+  static void PrepareBias(const float *input, float *output, Index rows, Index cols) {
```

Comment:
We should be using marian tensors here, as they allow for zero copy.

Reply:
To be more precise, when being inside the marian tensor ecosystem, you can choose to return the same unmodified tensor from prepareBias (identity operation that just returns the input tensor) and avoid the copy.

Reply:
`ruy_adapter.h` sits at the same level as `intgemm`. `Tensor` in arguments is therefore unsuitable. Marian _provides_ `Tensor`. Marian _uses_ `ruy_adapter` / `intgemm`.  My preferred solution to this is to capture the variations as callback/first-class functions and do f(args) to not duplicate code. These primitives should be supplied by this file/layer, without pulling in `marian::Tensor`.

Reply:
What I mean is that this function should never be called from ruy as it will be slower than the other code path. It should abort. The "shifted" code path shouldn't be taken by this backend.

Reply:
As discussed on today's call, since ARM has 8-bit signed * signed, the relevant code path to follow is int8 (not int8shifted).

Reply:
[`49beb50` (#79)](https://github.com/browsermt/marian-dev/pull/79/commits/49beb50ba1935bc2f27b20ec07804b2ba7cb09d8). 

@kpu suggested lifting ruy path from `intgemm_interface.h` and adding at a level of gemm-provider. Much of the existing state is due to coming over from https://github.com/jerinphilip/MozIntGemm. This will involve writing some equivalent of `intgemm_interface` to complete the calls (i.e duplication). Please let me know if it's a go on this one. 

> I agree. I understand the objective here is to avoid the std::memcpy which can be done by changing the calling code. I will try to find a solution that achieves the intent at integration.

The std::memcpy was probably useless, got rid of it and a bunch of `nullptr` in the args in the process.

Reply:
> @kpu suggested lifting ruy path from intgemm_interface.h and adding at a level of gemm-provider. Much of the existing state is due to coming over from [jerinphilip/MozIntGemm](https://github.com/jerinphilip/MozIntGemm). This will involve writing some equivalent of intgemm_interface to complete the calls (i.e duplication). Please let me know if it's a go on this one.

Without documentation (for `intgemm_interface.h`, the integration to marian's graph system) or active help from someone who understands the area, this is unfortunately not within the immediate future.

Reply:
I have read https://github.com/browsermt/marian-dev/blob/master/src/tensors/cpu/intgemm_interface.h and determined that, while it could use some comments, it is in state where any software engineer working for me is expected to read though and understand or come up with questions.